Support ignoring parameters in FSDP2

ckluk2 opened the pull request:
Differential Revision: D69153051




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

awgu commented on torch/distributed/fsdp/_fully_shard/_fully_shard.py:
```diff
@@ -46,6 +48,7 @@ def fully_shard(
     shard_placement_fn: Optional[Callable[[nn.Parameter], Optional[Shard]]] = None,
     mp_policy: MixedPrecisionPolicy = MixedPrecisionPolicy(),
     offload_policy: OffloadPolicy = OffloadPolicy(),
+    ignored_params: set[nn.Parameter] = set(),
```
nit: I think it is a good practice to avoid mutable default args (https://docs.python-guide.org/writing/gotchas/) in case the function mutates it (which might not be the case here).
```suggestion
    ignored_params: Optional[set[nn.Parameter]] = None,
```

```
if ignored_params is None:
    ignored_params = set()
```

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_ignore_params.py:
```diff
@@ -0,0 +1,326 @@
+                                                    ignored_params=fsdp_ignored_params)
+
+        fsdped_modules = _find_all_fsdped_modules(test_model, "")
+        # print(f">>> [rank {dist.get_rank()}]: FSDPed modules: {fsdped_modules}")
```
clean up the prints before landing

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_ignore_params.py:
```diff
@@ -0,0 +1,326 @@
+import torch.nn as nn
+
+from torch.distributed._composable.fsdp import fully_shard
+
```
nit: I wonder if the OSS linter would compact these. Maybe check `lintrunner`.

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_ignore_params.py:
```diff
@@ -0,0 +1,326 @@
+    return modified_list
+
+
+def _print_all_params(title: str, name_to_param_map: dict):
+    msg = f"{title}:\n"
+    for name, param in sorted(name_to_param_map.items()):
+        msg += f"name={name}, param={param}\n"
+    print(msg)
+
+
+def _print_param_set(title: str, param_set: set[torch.nn.Parameter], param_to_name_map):
+    msg = f">>> [rank {dist.get_rank()}]: {title}"
+    msg += "{"
+    for param in param_set:
+        assert param in param_to_name_map
+        name = param_to_name_map[param]
+        msg += f"{name}, "
+    msg += "}"
+    print(msg)
```
nit: Are these for debugging? Do you want to keep them in the unit test when landgin?

awgu reviewed (commented):
Nice! The FSDP logic changes look good to me, and thanks for the thorough testing.

Let's do one round of cleanup, and I think we can land. Let us also let @weifengpy to review to have the context.

weifengpy commented on torch/distributed/fsdp/_fully_shard/_fsdp_init.py:
```diff
@@ -70,7 +72,51 @@ def _get_device_from_mesh(mesh: DeviceMesh) -> torch.device:
+            return False
+
+    # Need to consider descendants of module
+    for _, child in list(module.named_children()):
```
is `named_children` necessary here? or `module.children()` is enough?

weifengpy commented on test/distributed/_composable/fsdp/test_fully_shard_ignore_params.py:
```diff
@@ -0,0 +1,326 @@
+        ignored_path = "module_b.module_c"
+
+        fsdp_ignored_params = _discover_fsdp_ignored_params(test_model, ignored_path=ignored_path, path="")
+        _print_param_set("fsdp_ignored_params:", fsdp_ignored_params, test_param_to_name_map)
```
ditto: replace printing with assert before landing

weifengpy reviewed (commented):
thanks for doing this! left minor comments

ckluk2 committed:
Support ignoring parameters in FSDP2 (#146631)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/146631

Test Plan: scripts/ckluk/fsdp/test_fully_shard_ignore_params.sh

Differential Revision: D69153051

huydhn commented:
@pytorchbot drci

huydhn commented:
@pytorchbot merge -f 'Landed internally'
